data/scenarios/Challenges/Ranching/gated-paddock/enclosure-checking.sw

```diff
@@ -0,0 +1,87 @@
+// Algorithm:
```

Comment:
How about requiring that _the fence must not branch and must connect the north and south solid fence_?

That would simplify the checking a lot, just follow the path using a scanner. Then at least one sheep has to be surrounded by the fence:
```
 +------+
++######|
|## ^  #v
|#  @  /
|#  v  #^
|#######|
+---|---+
    ^
start here and go left and right
```

So if you go north, find a fence and go left, you must reach the south solid fence and if you go right the north solid fence.
Similarly from the fence that must be south of the sheep.

This check will short-circuit a lot faster if the fence is not connected or a sheep can just go south or north to the ocean.

Reply:
More simply (perhaps), just (1) start at the north gate and follow fence until reaching the south gate (making sure it does not branch and there are no gaps), then (2) start at a sheep and go in a straight line until reaching the ocean, counting fence+gates encountered (being careful to count contiguous sections as only one).  If a fence was encountered an odd number of times, the sheep is inside.

Reply:
@byorgey but if you add any other line of fence unconnected to the gate, your check is not valid, right? 🤔 

I did not want to check all island locations for unconnected fences, so going south and north seemed simpler.

Reply:
Oh, good point.  You could somehow mark every fence section you traversed while walking along it, and then only count marked fence sections.

Reply:
One thing I'd intended but haven't stated yet is for the player to be able to incorporate natural barriers (e.g. mountains) into their fencing strategy.  For instance, the player might extend the fence a bit further than they would otherwise to reach a mountain range that, in the end, saves more fencing after all.

I hadn't demoed this yet because the robot program would be more complicated than just drawing a rectangular fence without obstacles.

Point being, whatever alternative algorithm is used to check for enclosure should account for non-fence barriers (of potentially multi-tile thickness.

Reply:
@kostmo I would suggest using more complicated checks in more complicated challenges.

Notice that chess horse does not implement full chess rules and not all mazes implement the general shortest path algorithm in an undirected graph.

Using solutions tailored to the specifics of the challenge leads to better performance and robots can `say` to the player what is wrong. Ideally, the Swarm code should also be maintainable, so that when things break other people can fix it.

---

What I am advocating for here is making the challenge easy for players and developers alike.

Small things like deciding that the fence has to connect to the starting gate improve the performance by a lot.
That in turn makes it unnecessary to add extra requirements like `say "Please check now"`.